dom.bs

```diff
@@ -5828,6 +5855,11 @@ dictionary ShadowRootInit {
   required ShadowRootMode mode;
   boolean delegatesFocus = false;
 };
+
+dictionary GetInnerHTMLOptions {
+  boolean includeShadowRoots = true;
+  sequence&lt;ShadowRoot> closedRoots;
```

Comment:
I continue to find this API shape problematic. We really want to have a consistent API surface. Given the [current consensus](https://github.com/WICG/webcomponents/issues/79#issuecomment-432974389) for `getComposedRange` for `Selection` is that it takes a list of shadow roots regardless of their closed-ness, we should match that convention and always take a list of shadow roots here as well.

Reply:
Hmm, I only have the context from the issue you linked, and not directly, but I assume you're referring to this:

> Rough consensus at TPAC F2F:
> 
> * Add `getComposedRange`, which takes a list of closed trees, to [`Selection`](https://w3c.github.io/selection-api/#dom-selection).

I can't help but notice the "which takes a list of **closed trees**" there...? Also, that thread seems dead for 2.5 years. Perhaps **this** could be the consistent API surface, which selection tries to match, once the discussion restarts?

Reply:
To match the selection API, this should be instead instead called getComposedInnerHTML and simply take an array of closed shadow roots.

I really don't think we should add this one off API before we can reach a better consensus on what the general approach on how to define all these APIs for the composed tree & flat tree. We need to stop introducing these one-off APIs with slightly different semantics. We've done enough damages to the web platform already. We need to stop.

Beyond that, I have some issues with the semantics of how serialization is proposed. I'd add a new comment on https://github.com/whatwg/html/pull/5465/files.

Reply:
Sorry - can you link me to the spec for `getComposedInnerHTML`? I can't seem to find that one. Perhaps you're just referring to the [unimplemented 2018 proposal](https://github.com/WICG/webcomponents/issues/79#issuecomment-432974389) for `getComposedRange`?

This seems different from that. The word "composed" I'd even object to here. We're just trying to return the full HTML for a node, including shadow content. Since we can't break the web by starting to do that with `innerHTML`, we need a new API. The closest thing to the existing `innerHTML` seems to be `getInnerHTML()` which allows the introduction of parameters, and is easily understandable by developers. Renaming this to `getComposedInnerHTML` will make everyone go look up "composed" to try to see if this'll just return them the HTML, which it will. So we make it longer to type, and more confusing to developers. Can you help me understand what we gain in the bargain?

Thanks for your comments on https://github.com/whatwg/html/pull/5465/files. I've put some replies there.

Reply:
So that people don't accidentally use this variant instead of innerHTML. We really shouldn't be encouraging people to serialize contents of shadow roots. Just like `Element.prototype.shadowRoot` should only be used when it's absolutely necessary, this variant of innerHTML should only be used when it's absolutely necessary.

Reply:
I had also raised this in https://github.com/mfreed7/declarative-shadow-dom/issues/9.

Reply:
Please see my comment in https://github.com/mfreed7/declarative-shadow-dom/issues/9#issuecomment-809793894. I've checked with the early implementers (e.g. Origin Trial participants) and they would be ok with these changes. But I don't want to allow too much time for people to start making use of the `getInnerHTML()` API such that we cause compat issues by changing the API shape.